Source/Core/DolphinQt/Debugger/CodeDiffDialog.cpp

```diff
@@ -0,0 +1,383 @@
+          &CodeDiffDialog::OnContextMenu);
+}
+
+void CodeDiffDialog::ClearData()
+{
+  if (m_record_btn->isChecked())
+    m_record_btn->toggle();
+  ClearBlockCache();
+  m_matching_results_list->clear();
+  m_exclude_size_label->setText(QString::fromStdString("Excluded: 0"));
+  m_include_size_label->setText(QString::fromStdString("Included: 0"));
+  m_exclude_btn->setEnabled(false);
+  m_include_btn->setEnabled(false);
+  std::vector<Diff>().swap(m_include);
+  std::vector<Diff>().swap(m_exclude);
+  JitInterface::SetProfilingState(JitInterface::ProfilingState::Disabled);
+}
+
+void CodeDiffDialog::ClearBlockCache()
+{
+  Core::State old_state = Core::GetState();
+
+  if (old_state == Core::State::Running)
+    Core::SetState(Core::State::Paused);
+
+  JitInterface::ClearCache();
+
+  if (old_state == Core::State::Running)
+    Core::SetState(Core::State::Running);
+}
+
+void CodeDiffDialog::OnRecord(bool enabled)
+{
+  JitInterface::ProfilingState state;
+
+  if (enabled)
+  {
+    ClearBlockCache();
+    m_record_btn->setText(tr("Recording..."));
+    state = JitInterface::ProfilingState::Enabled;
+    m_exclude_btn->setEnabled(true);
+    m_include_btn->setEnabled(true);
+  }
+  else
+  {
+    ClearBlockCache();
+    m_record_btn->setText(tr("Start Recording"));
+    state = JitInterface::ProfilingState::Disabled;
+    m_exclude_btn->setEnabled(false);
+    m_include_btn->setEnabled(false);
+  }
+
+  m_record_btn->update();
+  JitInterface::SetProfilingState(state);
+}
+
+void CodeDiffDialog::OnIncludeExclude(bool include)
+{
+  bool isize = m_include.size() != 0;
+  bool xsize = m_exclude.size() != 0;
+  std::vector<Diff> current_diff;
+  Profiler::ProfileStats prof_stats;
+  auto& blockstats = prof_stats.block_stats;
+  JitInterface::GetProfileResults(&prof_stats);
+  std::vector<Diff> current;
+  current.reserve(20000);
+
+  // Convert blockstats to smaller struct Diff. Exclude repeat functions via symbols.
+  for (auto& iter : blockstats)
+  {
+    Diff tmp_diff;
+    std::string symbol = g_symbolDB.GetDescription(iter.addr);
+    if (!std::any_of(current.begin(), current.end(),
+                     [&symbol](Diff& v) { return v.symbol == symbol; }))
+    {
+      tmp_diff.symbol = symbol;
+      tmp_diff.addr = iter.addr;
+      tmp_diff.hits = iter.run_count;
+      current.push_back(tmp_diff);
+    }
+  }
+
+  // Could add address based difference instead of symbols. Probably need second function.
+  // Sort for lower_bound.
+  sort(current.begin(), current.end(),
+       [](const Diff& v1, const Diff& v2) { return (v1.symbol < v2.symbol); });
+
+  // If both lists are empty, write and skip.
+  if (!isize && !xsize)
+  {
+    if (include)
+      m_include = current;
+    else
+      m_exclude = current;
+    return;
+  }
+
+  // Update exclude list. Compare to exclude list if it exists and make current_diff to check
+  // against include.
+  if (!xsize && !include)

```

Comment:
Not sure how feasible it is but I find it a little hard to follow the code here due to all the various checks for include or exclude along with the container checks for empty or not.  It might be helpful to split this into multiple functions, one for each capability, ex:

```
if (include)
{
  OnIncludeLogic();
}
else
{
  OnExcludeLogic();
}

// Do anything common...
```

Reply:
Looking into this now

Reply:
After re-reviewing the existing code we would end up with a lot of repeated code if this is split up. The common steps are sprinkled in between non-common steps and are context sensitive.

I would opt to not change currently. During the rewrite I suspect a lot of this code will be tossed out.

Reply:
I wrote out the code into separate functions, removing pieces of logic where I knew `included` was false or true in the given function.  This is what I ended up with:

```


void OnInclude()
{
	std::vector<Diff> current_diff;
	auto current_symbols = CalculateSymbolsFromProfile();
	if (m_include.empty() && m_exclude.empty())
	{
		m_include = current_symbols;
		return;
	}
	
	if (m_exclude.empty())
	{
		current_diff = current_symbols;
	}
	else
	{
		for (auto& iter : current)
		{
			auto pos = lower_bound(m_exclude.begin(), m_exclude.end(), iter.symbol, AddrOP);

			if (pos->symbol != iter.symbol)
			{
				current_diff.push_back(iter);
			}
		}
	}
	
	if (!m_include.empty())
	{
		// Compare include with current and remove items that aren't in both.
		std::vector<Diff> tmp_swap;
		for (auto& iter : m_include)
		{
			if (std::any_of(current_diff.begin(), current_diff.end(),
					  [&](Diff& v) { return v.symbol == iter.symbol; }))
			{
				tmp_swap.push_back(iter);
			}
		}
		m_include.swap(tmp_swap);
	}
	else
	{
		 m_include = current;
		 RemoveMatchingSymbolsFromIncludes(m_exclude);
	}
}

void OnExclude()
{
	std::vector<Diff> current_diff;
	auto current_symbols = CalculateSymbolsFromProfile();
	if (m_include.empty() && m_exclude.empty())
	{
		m_exclude = current_symbols;
		return;
	}
	
	 if (m_exclude.empty())
	{
		m_exclude = current_symbols;
	}
	else
	{
		for (auto& iter : current_symbols)
		{
			auto pos = lower_bound(m_exclude.begin(), m_exclude.end(), iter.symbol, AddrOP);

			if (pos->symbol != iter.symbol)
			{
				current_diff.push_back(iter);

				m_exclude.insert(pos, iter);
			}
		}
		// If there is no include list, we're done.
		if (m_include.empty())
			return;
	}
	
	if (m_exclude.empty())
	{
		RemoveMatchingSymbolsFromIncludes(m_exclude);
	}
	else
	{
		RemoveMatchingSymbolsFromIncludes(current_diff);
	}
}

std::vector<Diff> CalculateSymbolsFromProfile()
{
	Profiler::ProfileStats prof_stats;
	auto& blockstats = prof_stats.block_stats;
	JitInterface::GetProfileResults(&prof_stats);
	std::vector<Diff> current;
	current.reserve(20000);

	// Convert blockstats to smaller struct Diff. Exclude repeat functions via symbols.
	for (auto& iter : blockstats)
	{
		Diff tmp_diff;
		std::string symbol = g_symbolDB.GetDescription(iter.addr);
		if (!std::any_of(current.begin(), current.end(),
						 [&symbol](Diff& v) { return v.symbol == symbol; }))
		{
			tmp_diff.symbol = symbol;
			tmp_diff.addr = iter.addr;
			tmp_diff.hits = iter.run_count;
			current.push_back(tmp_diff);
		}
	}

	// Could add address based difference instead of symbols. Probably need second function.
	// Sort for lower_bound.
	sort(current.begin(), current.end(),
		[](const Diff& v1, const Diff& v2) { return (v1.symbol < v2.symbol); });
		
	return current;
}

void RemoveMatchingSymbolsFromIncludes(const std::vector<>& symbol_list)
{
	for (auto& list : symbol_list)
	{
		m_include.erase(std::remove_if(m_include.begin(), m_include.end(),
									 [&](Diff& v) { return v.symbol == list.symbol; }),
					  m_include.end());
	}
}
```

Assuming it works, I think it's much clearer to understand.

it also makes it a bit more obvious where there's some odd logic.  Like for instance, this line:

```
else if ((m_include.empty() && include) || (m_exclude.empty() && !include))
```

which makes no sense.  If `m_exclude` is empty, then the corresponding loop will never be performed, so the check doesn't really work imo.

Reply:
I'm fine with it being changed.  As for that weird logic case, I have no idea.
Trying to integrate what you wrote- replaced OnIncludeExclude and added the bool check only in Update() to call the new functions.


The signature here:
`void RemoveMatchingSymbolsFromIncludes(const std::vector<>& symbol_list)`
is missing a type contained in the vector. What was the reason for it to be a const?

Reply:
On
`m_include = current;`

Do you intend that current to be the same current in `CalculateSymbolsFromProfile`?
Should it become a member variable? Right now `current` is constrained to `CalculateSymbolsFromProfile` only.

Reply:
I went ahead and implemented your change and did a bunch of testing with it to make sure the logic stays the same.
Great job, it's a lot cleaner too than before!

Reply:
Thanks @dreamsyntax !  Glad it worked out.  I'll give it another look over in a bit.  Appreciate all your work on this.

Reply:
@iwubcode is it recommended I squash this feature into one commit (assuming no further changes are needed and this actually would be merged)? Or can GitHub's web merge auto squash?

Reply:
@dreamsyntax - yeah usually, we try to limit the number of commits down to logical pieces.  Since this is all one set of changes, it probably makes most sense in a single commit.  You'll need to squash that before a merge can happen.